File: Data Structures & Algorithms/reorder-linked-list/submission-1.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        #divide list into two halves using slow/fast pointers
        slow = head
        fast = head
        while fast.next and fast.next.next:
            slow = slow.next
            fast = fast.next.next
        second_half = slow.next
        slow.next = None
        first_half = head

        cur1 = first_half
        cur2 = second_half

        #reverse the second half
        prev = None
        curr = second_half
        while curr:
            nxt = curr.next
            curr.next = prev
            prev = curr
            curr = nxt
        second_half = prev

        #merge two lists
        cur1 = head
        cur2 = second_half
        while cur2:
            nxt1 = cur1.next
            nxt2 = cur2.next
            cur1.next = cur2
            cur2.next = nxt1
            cur1 = nxt1
            cur2 = nxt2

File: Data Structures & Algorithms/merge-k-sorted-linked-lists/submission-1.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        if (lists.empty())
            return nullptr;
        while (static_cast<int>(lists.size()) > 1) {
            vector<ListNode*> merged;
            for (int i = 0; i < static_cast<int>(lists.size()); i += 2) {
                ListNode* l1 = lists[i];
                ListNode* l2 = ((i + 1) < static_cast<int>(lists.size())) ? lists[i+1] : nullptr;
                merged.push_back(mergeTwo(l1, l2));
            }
            lists = std::move(merged);
        }
        return lists[0];
    }

private:
    ListNode* mergeTwo(ListNode* l1, ListNode* l2) {
        ListNode dummy(0);
        ListNode* curr = &dummy;

        while (l1 && l2) { 
            if (l1->val <= l2->val) {
                curr->next = l1;
                l1 = l1->next;
            } else {
                curr->next = l2;
                l2 = l2->next;
            }
            curr = curr->next;
        }
        curr->next = l1 ? l1 : l2;
        return dummy.next;
}
};
